### ISSUE TITLE
Watch Command

### ISSUE THREAD

**Issue by tehpsalmist** [ZengineHQ/mayan#11]
@alexweber mentioned that this topic has been discussed elsewhere, but I couldn't find those conversations, so here's a new issue.

I have a script that I have been using to watch my frontend files and automatically deploy on save. It works great for my purposes.

[This is it.](https://gist.github.com/tehpsalmist/fee4b6b5d22ec366381092f7b47b88db)

Alex and I also discussed the issue of rate of save/multiple concurrent file saves, which *might* warrant some sort of debounce effect on the save event, to avoid colliding deploy commands or hitting the API above the rate-limit. I think the more critical issue there would be to prevent the deploy command from executing while another deploy command is in action. As long as only one deploy executes at a time, that should avoid nasty errors.

I have really enjoyed working with the angular-cli (an incredible tool that has improved vastly in the last few months), so I am going to investigate their source code to see how they handle debouncing/throttling, and maybe get some ideas from that.

**Comment by tehpsalmist** [comment]
Obviously, what I'm currently doing by spinning up a new shell and executing the deploy command with child_process.exec won't be necessary from inside mayan, because we can just execute the javascript, so that's nice.

**Comment by evert0n** [comment]
I also think it's worth to look into a `watch` implementation with some throttling to avoid nuking the API
I have used `nodemon` with maya before for this

**Comment by tehpsalmist** [comment]
Have you run into particular scenarios using nodemon or other setups where you noticed the API was getting 'nuked'?

**Comment by tedkulp** [comment]
Just leaving this note here for later. I did try the nodemon this earlier today with the delay option and it works pretty well.  The command was something like:

`nodemon --verbose --watch "frontend" -e js,css,scss,json --ignore frontend/build --delay 10000ms --exec "mayan deploy frontend"`

Once in awhile, nodemon ignores the `--ignore` option and you have to restart it, but otherwise it works really well.

**Comment by tehpsalmist** [comment]
Woah, didn't know about that exec command. That's awesome.
